### PR TITLE
Fix #7741: Failure to report escaping type variable

### LINKED ISSUE #7741: Failure to report escaping type variable.
**Original bug ID:** 7741
**Reporter:** @Drup
**Status:** new
**Resolution:** open
**Priority:** normal
**Severity:** minor
**Category:** typing
**Related to:** #6744
**Monitored by:** @Drup @gasche

## Bug description

The attached file fails with 

   The implementation blurb.ml
       does not match the interface (inferred signature):
       Values do not match:
         val x : 'a bla t
       is not included in
         val x : 'a bla t

Regardless of how confusing that error message is, I'm pretty sure this code should fail with an error of the type "The universal variable 'a would escape its scope". 

I'm not sure you can break soudness like that, but it seems very sketchy.

Similar to ocaml/ocaml#6744

## File attachments

- [blurb.ml](https://gist.githubusercontent.com/vicuna/a20102e4996056b34c121034ae0fd16d/raw/a7378981a91b183e31f7b8b82f730e51eb78daa4/blurb.ml)

### PR DESCRIPTION
**Pull request by garrigue**
This is a fix for #7741.
Forgot to check for the `Inj` flag in `Ctype.occur_univar`.

### CODE REVIEW

**Comment by xavierleroy** [comment]
This is a bug fix.  Please review.

**Comment by gasche** [comment]
Okay, so beginner questions:

1. Is there some documentation somewhere of what `May_pos` and `May_neg` mean?
2. Why is this piece of code not traversing parameters unless they are `May_pos` or `May_neg`?
    (Maybe a comment could be nice?)
3. Maybe there is a different way to write this code that is less bug-prone, being of the form `if  <cases that are known to be special> then (be clever) else (do the safe thing)`, rather than `if (cases that are currently known to require a check) then (do the safe thing) else (be clever)`?
3. There are several other places in code that also check for `May_pos || May_neg` only; do they also need to be updated?

**Comment by garrigue** [comment]
@gasche Thanks for your comments!
In particular 4, as I had missed another occurrence of exactly the same code just a few lines below...
This is now fixed (identically). The other occurrences are only about variance inference itself, so there is no need to change them.

Concerning your other comments:
1. Variance flags are described in my OCaml 2013 presentation: [On variance, injectivity, and abstraction](http://www.math.nagoya-u.ac.jp/~garrigue/papers/index.html)
2. I added a comment.
3. As I wrote in the last line of the comment, the formula itself is always true, except on some public abbreviations. Namely, concrete types are injective, and abstract or private types are either marked with `+` or `-`, so at least one of `May_pos` or `May_neg` is true. This said, it could make sense to add the condition
```ocaml
if td.type_private = Private || td.type_kind <> Type_abstract then raise Not_found
```
Do you think it would make the code clearer?

**Comment by gasche** [comment]
Thanks for the clarifications.

I think we should have documentation for the 7 variances in the codebase itself, rather than in a PDF on your website. This can be delayed to another PR, but I think that delaying it indefinitely is not good.

The comment you added says:

> Optimization: rather than expanding the type definition
> for abbreviations, use the inferred presence information.

If I understand correctly, the code never expands the definition. What happens is that we know, for certain variances, that the type parameter cannot actually occur in the expanded type, and for those we don't need an occur check. (I thought you used a related strategy when fixing levels, but I don't know where that would be in the code.) For the parameters that may occur in the definition, we do the occur check.

I wish your comment would explain this: "if the variance is <...>, then we know the parameter does not occur in the type definition, so we can skip the occurrence check in the parameter".

There is a variance `Variance.null` whose documentation says "no occurrence". What is the difference between the check you are doing and `Variance.subset v null`?

**Comment by gasche** [comment]
> This said, it could make sense to add the condition [...]

In general I would rather have *less* code if we know that it is not necessary; so no, I would not be in favor of the addition.

File: typing/ctype.ml [hunk 1]
```diff
@@ -1913,7 +1913,10 @@ let occur_univar env ty =
             let td = Env.find_type p env in
             List.iter2
               (fun t v ->
-                if Variance.(mem May_pos v || mem May_neg v)
+                (* Optimization: rather than expanding the type definition
+                   for abbreviations, use the inferred presence information.
+                   Only public abbreviations may make this formula false. *)
```
Comment (Octachron) [inline typing/ctype.ml]: If we had bivariant annotation on abstract types, would they not make the formula false too?

**Comment by garrigue** [comment]
> I think we should have documentation for the 7 variances in the codebase itself, rather than in a PDF on your website. This can be delayed to another PR, but I think that delaying it indefinitely is not good.

You mean, in `types.mli`.
I'll do that (in another PR), but this doesn't replace the PDF, which is the basis for the computation `typedecl_variance.ml`.

> I wish your comment would explain this: "if the variance is <...>, then we know the parameter does not occur in the type definition, so we can skip the occurrence check in the parameter".
> 
> There is a variance `Variance.null` whose documentation says "no occurrence". What is the difference between the check you are doing and `Variance.subset v null`?

You're right, this is equivalent to `not Variance.(eq v null)`.
I'm going to change it that way. There is no need for an extra comment, as `types.mli` already says `no occurrence` for `null`.

**Comment by gasche** [comment]
Thanks! Now I understand the code, and I believe it is correct.

I am sorry to be nitpicky, but I'm not fond of the comment you added. I think that something along the lines that I had suggested would be better. You write:

> Optimization: rather than expanding the type definition
> for abbreviations, use the inferred presence information.
> Only public abbreviations may have a null variance.

To me, this comments suggests:
- That the optimization is to avoid the expansion of a type definition. But this function never expands type definitions anyway, it is only traversing the type expression. The optimization skips some subexpressions.
- That it is important to reason on whether the type may or may not be an abbrevation. But it is not; the optimization is correct for all types, abbreviations or not. (Yes, it can only occur on abbreviations, but why is this an important information here?)

If we want to have a comment here, it should explain what the intent of the code is, and why it is correct. Here would be a suggestion:

> Parameters with a `null` variance cannot occur in the definition of the type constructor.
> As an optimization, we skip the occur check on their instances.

Of course this is not perfect; in particular it does not explain why this is sound. I think that correctness depends on the interplay with unification and generalization. (For example, type parameters that do not occur in a *datatype* declaration cannot skip the occur check: they are injective and they can be extracted using a constraint for example.)

(I'm not sure why this is sound, but I am sure that the PR as implemented is not a regression compared to the previous state of the codebase, which I guess is enough to accept it.)

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
Another attempt with more components from your comment:

> The `null` variance only occurs in type abbreviations and
> corresponds to type variables that do not occur in the definition
> (expanding abbreviations would erase them completely). 
> The type-checker consistently ignores type expressions in this position.

**Comment by garrigue** [comment]
Changed comment as suggested.

**Comment by gasche** [comment]
Unless someone else volunteers to do this, I can merge and cherry-pick in 4.11 when the CI is happy.

**Comment by lpw25** [comment]
As I mentioned on #7741: It is not clear to me why we are not expanding the type abbreviations in this case. In the normal occurs check and the update levels check we expand abbreviations if it is needed to prevent something from escaping its scope. Here instead we use variance to check that expanding the abbreviation would fix the problem and then leave it unexpanded. I would expect this to cause univars to be able to escape their scope, however I have not been able to produce such an example.

I think one of the following is true, but I cannot determine which:
1. Univars can escape their scope by being present in the type parameter of a "bivariant" type alias.
2. Occurs_univar is not really needed because whenever a univar escapes its scope by unifying with a variable then the escape is detected by some other mechanism.
3. Occurs_univar is needed but whenever a univar escapes its scope via a "bivariant" type alias that alias is expanded via some other mechanism preventing the escape.

**Comment by gasche** [comment]
I would agree with @lpw25 that it is strange to not use the same strategy in all those scenarios (or at least it should be explained). Is there a good reason why we are not expanding all the time (for vanishing parameters) or not-expanding all the time?

(Another strategy might be to simply erase those type subexpressions when first encountering or creating them, replacing them with a dummy type -- unifying dummy with anything is dummy.)

**Comment by garrigue** [comment]
The answer is that it does escape, but you cannot observe it because this variable will consistently be ignored.
```ocaml
type 'a t = int
let f (x : < m : 'p. 'p t >) = (x : < m : 'a >)#m;;
val f : < m : 'p. 'p t > -> 'p t = <fun>
let f (x : < m : 'p. 'p t >) = x#m;;
val f : < m : 'p. 'p t > -> 'a t = <fun>
```
Whether we should hide it or not is political :)

To be more precise the strategy applied in `occur` is more general (at least until we recompute exact variances in functor results), but the normal occur check never has to look under a `Tobject`, whereas `occur_univar` has to do that, which may require to expand huge types in case of mutual recursion, so for performance reasons it seems better to avoid any kind of expansion in `occur_univar`.

**Comment by gasche** [comment]
Leo: my reasoning for approving despite the uncertainty about the variance-only strategy was as follows:

> I'm not sure why this is sound, but I am sure that the PR as implemented is not a regression compared to the previous state of the codebase, which I guess is enough to accept it.

If you agree with this reasoning, I would propose to merge this PR. We can keep discussing the expand-or-not strategy.

@garrigue: thanks for the additional explanations

> occur_univar has to [look under a Tobject], which may require to expand huge types in case of mutual recursion, so for performance reasons it seems better to avoid any kind of expansion in occur_univar.

This is a great point, could we add it as a comment in the code?

> To be more precise the strategy applied in occur is more general (at least until we recompute exact variances in functor results)

If I understand correctly:
- Sometimes we compute variances that are stricter than they need to be (because we don't refine them after functor application); but never variances that are more permissive than they should be (this would be unsound)
- The `occur` strategy does not use variance, instead it backtracks if the check fails and tries to expand the definition; this is "more general" in the case of a parameter that has stricter variance than `null`, but yes is vanishing.

This leads to two natural questions (that I think are orthogonal to this PR):
- If we really feel good about the null-variance optimization, why are we not also using it in `occur`?
  (in `update_level` I understand that this is to preserve the order invariant between levels and subexpressions depths)
- Maybe one could actually try the expansion-on-failure strategy in `occur_univar`, and see whether the performance on object-oriented codebases really suffers? (I would assume that it is fairly uncommon for this check to fail. The only place I see where it is not an error is `split_univars` in `unify_row_field`, and in that case a pessimistic variant that does not expand on failure would probably be fine?)

Note: "recomputing variances at functor application" can be expressed nicely by having "variance variables" implicit in the functor definition (for the variances of the abstract types of functor inputs), and "open variance expressions" (mentioning variance variables) in the type definitions in the functor result. (This hypothetical design shows that the feature preserves abstraction/composability, which was not completely obvious to me at first.)

**Comment by garrigue** [comment]
Where do we stand here?
@lpw25 do you still have a problem with this (incremental) fix.

**Comment by lpw25** [comment]
I have no problem with merging this fix.

I remain mildly sad about the state of `occur_univar`: I'm fairly adamant that we shouldn't allow things to escape from their scope even if "you cannot observe it because this variable will consistently be ignored". Not keeping your types well-formed is a recipe for subtle bugs and is completely unnecessary.

Given the other issues around `enter_poly`/`occur_univar` from the polymophic types discussion, I'm now planning to rewrite this part of the system to avoid needing `occur_univar` in the first place. Which will save us from needing to reach agreement about how it should behave.

**Comment by garrigue** [comment]
Thanks. Merged and cherry-picked on 4.11 as 74ad349033

BTW, I'm curious what you have in mind for the scope of univars. I suppose it would mean having  scopes on unification variables that indicate which univars they can safely capture, but that looks like a pervasive change.

**Comment by lpw25** [comment]
I was thinking of something along the lines of:
1. Keep two type maps which map univars to univars when doing unification (or equality or etc.). One map for each side of the unification.
2. Keep the set of unified pairs of univars (same as `univar_pairs` now)
3. When you enter a poly node you:
       a) increase the current level
       b) for each bound univar create a corresponding fresh univar whose scope is the new current level
       c) add a mapping from each bound univar to the corresponding fresh univar in the appropriate type map
4. When you encounter a univar during unification look up its fresh version in the appropriate map
4. When you need to unify two univars:
       a) run update_level on them as usual
       b) if one has already been made equal to a different univar then error
       c) add them to the set of pairs of unified univars

Basically it is just doing a lazy substitution for a fresh variable when you open a binding. I believe that this correctly handles recursive types.

**Comment by gasche** [comment]
I understand all of that (nitpick: you write "for each bound univar", you mean each univar bound in this poly node, but at first I thought you meant each univar already in the map (we traversed its binder earlier), which didn't make full sense), except the "run update_level on them as usual", what is the usual here? I would expect that if you try to lower the level of a "fresh univar" below its binding level, you should get an escape error.

(Indeed, I don't see a problem with recursive types.)

**Comment by garrigue** [comment]
I see part of the idea. However, you have to reconcile it with the graph structure. And it seems difficult to use the same levels, are they are supposed to decrease inside types, not increase.

Also, if we want to improve on the current approach, we want to be able to unify unification variables with types containing univars. This should be possible thanks to some kind of levels, but would still require something similar to `occur_univar`, but allowing some safe univars, or yet another new scoping mechanism.

**Comment by lpw25** [comment]
> except the "run update_level on them as usual", what is the usual here? I would expect that if you try to lower the level of a "fresh univar" below its binding level, you should get an escape error.

By "as usual" I meant the calls to `update_level` that are already in the `unify` function. Trying to lower the level of a "fresh univar" below its binding level should indeed give an escape error. The call to `update_level` is what actually emits that error.

> However, you have to reconcile it with the graph structure.

I believe that it is reconciled with the graph structure. Or at least, I take that statement to mean that it needs to work correctly in the presence of recursive types, and I believe that it does.

> And it seems difficult to use the same levels, are they are supposed to decrease inside types, not increase.

I would like to avoid adding another level mechanism if possible.

My suggestion is intended to be roughly equivalent to removing the `occur_univar` check and then replacing `enter_poly` with something like:
```ocaml
let enter_poly univar_pairs ty1 univars1 ty2 univars2 f =
  (* Increase the current level *)
  begin_def ();
  (* Take instances of the two types, copying only those parts
     which contain free references to the univars *)
  let vars1, ty1_instance = instance_poly ty1 univars1 in
  let vars2, ty2_instance = instance_poly ty2 univars2 in
  (* Replace the fresh variables with univars and give them
     a scope to stop them escaping *)
  List.iter (fun var ->
    set_type_desc var Tunivar;
    set_scope var !current_level
  ) (vars1 @ vars2);
  (* Remember the current set of equated univars *)
  let old_univar_pairs = !univar_pairs in
  (* Run the function (e.g. unification) on the instances *)
  f univar_pairs ty1_instance ty2_instance
  (* Restore univar_pairs *)
  univar_pairs := old_univar_pairs;
  (* Lower the current level *)
  end_def ();
  (* Re-run the function (e.g. unification) on the original types *)
   f univar_pairs ty1 ty2;
  (* Restore univar_pairs *)
  univar_pairs := old_univar_pairs
```
The idea is that we:
- First take instances of the types. This duplicates the parts of the types that contain free references to the univars bound by this poly. The duplicated parts are at the new level and the univars themselves are given scopes to stop them escaping.
- Then we run unification on those instances. This will link the parts of the original types that do not contain free references to the univars. It will also check that the univars do not escape their scopes.
- Then we re-run unification on the original types. This will link the parts of the original types that do contain free references to the univars. Thanks to the first unification we do not need to check for the univars escaping during this second unification.

This mechanism only relies on our usual level mechanism, but it traverses the type multiple times and copies the parts of the type that refers to the univars. My proposed algorithm aims to behave equivalently to this one, but to avoid these additional traversals (especially the quadratic aspect of unifying twice).

The main subtlety when comparing the above algorithm with the one I proposed is the treatment of the nodes that contain references to the bound univars but are not themselves univars. In the above algorithm they are duplicated for the first unification, which means they are given the fresh level, whereas in what I proposed they are not duplicated and so their level is not changed. IIUC this is what @garrigue is getting at with his above comment. Since we are effectively substituting the bound univars for new ones with a fresh level we are breaking the level invariant of these nodes by not also duplicating them.

I can see two possible solutions:

1. As Jacques says we can do something like `occur_univars`. In this case I think that would mean that, when linking a unification variable to a type, we would traverse the type to see if it contained a univar that was being substituted. If the scope attached to that univar was later than the level of the unification variable then we would emit a scope escape error.

2. We could emulate the behaviour of the two pass algorithm by replacing the map from bound univars to fresh univars with a map from types to levels/scopes. This allows us to do essentially both unifications simultaneously -- you use the map to lookup what the level/scope would be on the first unification and use that to detect escaping univars, and then you link the unmapped types.

I'm leaning towards the second approach, because I find the relationship with the naive algorithm easier to understand.

**Comment by garrigue** [comment]
OK. Now I see how your two-pass algorithm would work, by separating escape detection from unification of the actual types (i.e. the first pass only works on copies).

But with the current level mechanism, I do not see how you can use `update_level` for escape detection without doing this copy. The nice thing with levels is that the information is naturally cached, so that a type will only be visited if its level is higher than the level it must be lowered to. But this means that you need the type to originally have an higher level, and only copy can do that.

The quadratic cost of the two-pass approach is not that bad, as it is only quadratic in the nesting of `Tpoly`'s, and only paths leading to univars are copied, so that duplicated unifications only concern those paths. However the copy itself is costly as it requires building a reverse graph of the type (this could probably be optimised if needed). And I'm a bit concerned with the interaction with recursive types, as the memoized expansions are forgotten on copy.

**Comment by lpw25** [comment]
> And I'm a bit concerned with the interaction with recursive types, as the memoized expansions are forgotten on copy.

You're talking about recursive type aliases right? I think those are fine because I don't think it is possible for the memoized expansions from a recursive type aliases to also have a free reference to a univar. That would require that the recursive type alias be non-regular. That means that such type constructors won't be copied so we don't need to worry about losing their expansions.

**Comment by garrigue** [comment]
Recursive occurrences of memoized expansions should indeed never be copied in your scheme. However, the expansion is also memoized in the original abbreviation node.
So, yes, I believe you won't enter an infinite loop. You may still have to redo expansions, which can be costly too.
Still, it seems worth experimenting with the two-pass approach, as it seems easy to implement correctly. The only case where it wouldn't work is when unifying types that are both at `generic_level`.
